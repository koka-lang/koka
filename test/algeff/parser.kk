// --------------------------------------------------------------
// Parser example
// --------------------------------------------------------------

// First defined the `many` effect to describe multiple of results and failure
effect many {
  fun flip() : bool
  fun fail() : a
}

fun select(xs) {
  match(xs) {
    Nil        -> fail()
    Cons(x,xx) -> if (flip()) then x else select(xx) 
  }
}

// Return all solutions 
val solutions = handler {
  return x -> [x]
  fail()   -> []
  flip()   -> resume(True) + resume(False)
}


// Return the first solution
val eager = handler {
  return x -> [x]
  fail()   -> []
  flip()   -> match(resume(True)) {
    Nil -> resume(False)
    xs  -> xs
  }
}

// Choice is now defined over `flip` to possibly return multiple results
fun choice(p1,p2) {
  if (flip()) then p1() else p2()
}

// Basic combinators
fun many(p : () -> <many,div|e> a ) : <div,many|e> list<a> {
  choice { many1(p) } { [] } 
}

fun many1(p : () -> <many,div|e> a ) : <div,many|e> list<a> {
  Cons(p(),many(p))
}


//---------------------------------------------------------------
// Now define the parse effect

effect parse {
  fun satisfy<b>( pred: string -> maybe<(b,string)> ) : b
}

alias mparse = <parse,many>

// The parse handler uses the `many` effect to return results
val parse : forall<a,e> (string, () -> <parse,many|e> a) -> <many|e> (a,string)
  = handler(input:string) {
      return x -> (x,input)
      satisfy(pred) -> {
        match(pred(input)) {
          Nothing          -> fail()
          Just((res,rest)) -> resume(res,rest)
        }
      }      
    }


fun char(p : char -> bool) : parse char {
  satisfy( fun(input) {
    val sc = input.head
    match (sc.list) {
      [c] | p(c) -> Just((c,input.tail))
      _   -> Nothing 
    }
  })
}

// Expression parsers
fun symbol(c : char) : parse () {
  char( fun(d) { c == d })
  ()
}

fun digit() : parse int {
  int(char(digit?) - '0')
}

fun alpha() : parse int {
  int(char(alpha?) - 'a')
}

fun number() {
  many1(digit).foldl(0, fun(n,d) { 10*n + d })
}

fun expr() : <div,mparse> int {
  choice { 
    val i :int = term()
    symbol('+')
    val j = term()
    i+j
  }
  { term() }
}

fun term() {
  choice {
    val i : int = factor()
    symbol('*')
    val j = factor()
    i*j
  }
  { factor() }
}

fun factor() {
  choice(number) {
    symbol('(')
    val i = expr()
    symbol(')')
    i
  }
}


// Testing
// Parse for all solutions
fun mparse(p,input) {
  solutions{ parse(input,p) }
}

// Parse eagerly
fun eparse(p,input) {
  eager{ parse(input,p) }
}

fun test1() {
  mparse( expr, "1+2*3").show-parse-int.println 
}

fun main() {
  test1()
}

// Some show functions
fun show-parse( xs : list<(a,string)>, f : a -> string ) : string {
  xs.show-list fun(r) {
    r.show-tuple(f,show)
  }
}

fun show-parse-int( xs : list<(int,string)> ) : string {
  xs.show-parse(show)
}

fun show-parse-ints( xs : list<(list<int>,string)> ) : string {
  xs.show-parse( fun(is) { is.show-list(show) } )
}
