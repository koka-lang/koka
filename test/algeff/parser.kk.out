[01;30moperator branch (eager) flip: resume scoped
[00;39;49m[00m[01;30moperator branch (eager) fail: resume never
[00;39;49m[00m[01;30moperator branch (parse) satisfy: resume tail
[00;39;49m[00m[01;30moperator branch (solutions) flip: resume scoped
[00;39;49m[00m[01;30moperator branch (solutions) fail: resume never
[00;39;49m[00m[(7,""),(3,"*3"),(1,"+2*3")]
.Op-fail : forall<a> .op-fail<a>
.Op-flip : .op-flip
.Op-satisfy : forall<a> (pred : (string) -> maybe<(a, string)>) -> .op-satisfy<a>
.Ops-fail : forall<a,b> (fail : .op-fail<b>) -> .ops-many<b>
.Ops-flip : forall<a> (flip : .op-flip) -> .ops-many<bool>
.Ops-satisfy: forall<a,b> (satisfy : .op-satisfy<b>) -> .ops-parse<b>
.tag-many : string
.tag-parse : string
alpha : () -> parse int
char : (p : (char) -> bool) -> parse char
choice : forall<a,e> (p1 : () -> <many|e> a, p2 : () -> <many|e> a) -> <many|e> a
digit : () -> parse int
eager : forall<a,e> (action : () -> <many|e> a) -> e list<a>
eparse : forall<a,e> (p : () -> <many,parse|e> a, input : string) -> e list<(a, string)>
expr : () -> <many,parse,div> int
factor : () -> <many,parse,div> int
fail : forall<a> () -> many a
flip : () -> many bool
main : () -> <div,console> ()
many : forall<a,e> (p : () -> <many,div|e> a) -> <many,div|e> list<a>
many1 : forall<a,e> (p : () -> <many,div|e> a) -> <many,div|e> list<a>
mparse : forall<a,e> (p : () -> <many,parse|e> a, input : string) -> e list<(a, string)>
number : () -> <many,parse,div> int
parse : forall<a,e> (string, () -> <parse,many|e> a) -> <many|e> (a, string)
satisfy : forall<a> (pred : (string) -> maybe<(a, string)>) -> parse a
select : forall<a> (list<a>) -> many a
show-parse : forall<a> (xs : list<(a, string)>, f : (a) -> string) -> string
show-parse-int: (xs : list<(int, string)>) -> string
show-parse-ints: (xs : list<(list<int>, string)>) -> string
solutions : forall<a,e> (action : () -> <many|e> a) -> e list<a>
symbol : (c : char) -> parse ()
term : () -> <many,parse,div> int
test1 : () -> <div,console> ()